### Programming Project

In [63]:
import pandas as pd
import numpy as np
np.random.rand(3,4)


array([[0.26472889, 0.92653579, 0.24456285, 0.98834941],
       [0.59432693, 0.18494013, 0.39865104, 0.72013553],
       [0.35574038, 0.55157278, 0.87603959, 0.10705935]])

In [8]:
import pandas as pd
import numpy as np

#df = pd.read_csv('CSV_file.csv')

size_of_sample = 10

#rows = np.random.choice(df.index.values, size_of_sample)

#df_sample = df.loc[rows]

In [2]:
import pandas as pd
import numpy as np
df = pd.read_csv('Calls.csv')
df.describe

UnicodeDecodeError: 'utf-8' codec can't decode byte 0x80 in position 13: invalid start byte

In [ ]:
size_of_sample = 10

rows = np.random.choice(df.index.values, size_of_sample)

df_sample = df.loc[rows]

In [ ]:
np.random.choice(5, 3, replace=False, p=[0.1, 0, 0.3, 0.6, 0])


## Date simulation !!!!!!!!!!!!!

In [41]:
rng = pd.date_range('1/1/2006', periods = 150, freq = 'M')
rng

DatetimeIndex(['2006-01-31', '2006-02-28', '2006-03-31', '2006-04-30',
               '2006-05-31', '2006-06-30', '2006-07-31', '2006-08-31',
               '2006-09-30', '2006-10-31',
               ...
               '2017-09-30', '2017-10-31', '2017-11-30', '2017-12-31',
               '2018-01-31', '2018-02-28', '2018-03-31', '2018-04-30',
               '2018-05-31', '2018-06-30'],
              dtype='datetime64[ns]', length=150, freq='M')

## Generate random Call Centre data with np.random.randn using dates from rng

In [165]:
ts = pd.DataFrame(np.random.poisson(6, len(rng)), index=rng, columns = ["Cardiac"])
ts["Respitory"] = pd.DataFrame.assign(np.random.poisson(4, len(rng)))
ts["Strokel"] = pd.DataFrame.assign(np.random.poisson(1, len(rng)))
ts["Alcahol"] = pd.DataFrame.assign(np.random.poisson(10, len(rng)))
ts

Cardiac  Respitory  Strokel  Alcahol
2006-01-31        7          4        1       12
2006-02-28        7          2        1       12
2006-03-31        5          2        2        8
2006-04-30        8          1        1        7
2006-05-31        6          2        2       14
2006-06-30        6          4        2        7
2006-07-31        4          2        0        8
2006-08-31        6          5        4        7
2006-09-30        8          6        2       10
2006-10-31        8          3        2       14
2006-11-30        4          2        0        9
2006-12-31       11          3        1       12
2007-01-31        6          5        1        9
2007-02-28        4          4        3        9
2007-03-31        8          6        1       12
2007-04-30       11          5        0        5
2007-05-31        9          8        0       11
2007-06-30        4          5        3       11
2007-07-31        5          1        2        8
2007-08-31        7          4        1       10
2007-09-30        5          5        0        9
2007-10-31        5          5        0        6
2007-11-30        9          4        2       14
2007-12-31        3          5        0        8
2008-01-31        8          6        0       11
2008-02-29        8          1        2       12
2008-03-31       11          2        0        9
2008-04-30        5          4        1       12
2008-05-31        4          5        2       12
2008-06-30        8          0        1       11
...             ...        ...      ...      ...
2016-01-31        8          5        1       13
2016-02-29       10          3        0        9
2016-03-31        4          7        0       14
2016-04-30        5          2        2       12
2016-05-31        7          1        0        9
2016-06-30        4          3        1        8
2016-07-31        9          0        1        8
2016-08-31        6          4        0       17
2016-09-30       12          2        1        6
2016-10-31        7          3        0        9
2016-11-30        4         11        1       15
2016-12-31        5          1        1       11
2017-01-31        5          4        1       15
2017-02-28        7          1        0       12
2017-03-31        4          4        1        9
2017-04-30        5          3        0       13
2017-05-31        6          4        1       15
2017-06-30       10          4        0       10
2017-07-31        2          4        0       12
2017-08-31        6          2        1        8
2017-09-30        6          0        1        9
2017-10-31       11          9        1       11
2017-11-30        8          4        2       10
2017-12-31        4          2        0        8
2018-01-31        7          3        2       11
2018-02-28        3          5        1       10
2018-03-31        1          1        0        4
2018-04-30        4          7        4       10
2018-05-31       12          3        0       14
2018-06-30        4          4        2        6

[150 rows x 4 columns]

In [167]:
ts.mean()

Cardiac       6.053333
Respitory     3.846667
Strokel       0.920000
Alcahol      10.260000
dtype: float64

In [168]:
ts.describe()

Cardiac   Respitory     Strokel     Alcahol
count  150.000000  150.000000  150.000000  150.000000
mean     6.053333    3.846667    0.920000   10.260000
std      2.600430    2.025796    0.916076    2.981892
min      0.000000    0.000000    0.000000    4.000000
25%      4.000000    2.000000    0.000000    8.000000
50%      6.000000    4.000000    1.000000   10.000000
75%      8.000000    5.000000    1.000000   12.000000
max     14.000000   11.000000    4.000000   18.000000

In [174]:
ts.loc['2017-11-30':'2018-05-31']

Cardiac  Respitory  Strokel  Alcahol
2017-11-30        8          4        2       10
2017-12-31        4          2        0        8
2018-01-31        7          3        2       11
2018-02-28        3          5        1       10
2018-03-31        1          1        0        4
2018-04-30        4          7        4       10
2018-05-31       12          3        0       14

In [173]:
ts.loc['2017-11-30':'2018-05-31'].mean()

Cardiac      5.571429
Respitory    3.571429
Strokel      1.285714
Alcahol      9.571429
dtype: float64

In [175]:
ts.resample('Y').mean()

Cardiac  Respitory   Strokel    Alcahol
2006-12-31  6.666667   3.000000  1.500000  10.000000
2007-12-31  6.333333   4.750000  1.083333   9.333333
2008-12-31  6.416667   3.583333  1.083333  10.916667
2009-12-31  6.250000   3.333333  1.000000  10.583333
2010-12-31  5.833333   4.416667  0.750000   9.166667
2011-12-31  5.000000   2.833333  0.750000   9.916667
2012-12-31  5.333333   3.500000  1.000000  10.416667
2013-12-31  5.750000   3.500000  0.833333  11.166667
2014-12-31  7.166667   5.333333  0.833333  10.000000
2015-12-31  5.416667   5.000000  0.583333  10.250000
2016-12-31  6.750000   3.500000  0.666667  10.916667
2017-12-31  6.166667   3.416667  0.666667  11.000000
2018-12-31  5.166667   3.833333  1.500000   9.166667

### Poisson distribution
The Poisson distribution, named after French mathematician S. D. Poisson, is a discrete probability distribution that expresses the likelihood of a given number of events occurring in a fixed interval of time (and/or space) if these events occur with a known average rate and independently of the time since the last event occurred.

___

### Abandon Rate
Abandon rate is the number of calls that hang-up before connecting to an agent. This number does not include those calls that receive a busy signal.
Calculation: Abandoned Calls / Total Incoming calls
When analyzing your abandon rate, be aware that there are a small percentage of people who call the wrong number and hang-up once they hear your company name on the automated message (i.e.: "Thank you for calling ABC Corporation."). Typically these false abandon rates are calculated within the first 10 seconds and in most centres this can be between 1-2% of all total calls. This number may increase if a toll-free number is similar to another that receives high call volume.
Typically, abandon rates are linked to how fast call centre agents answer calls. The faster a call is answered, the lower the abandon rate. High abandon rates can result in lost sales opportunities and poor customer service. High abandon rates have the potential to artificially inflate future call volume as the initial customers who could not get through on the first call continue to call back until they reach an agent.
Example:
You receive 100 calls in one hour. Of those calls, 10 people hang up before speaking to an agent, so your abandon rate for the last hour has been 10%.
### Average Talk Time (ATT) & After Call Work (ACW) & Average Handle Time(AHT)
Average-Talk-Time (ATT) is the average amount of time agents talk to customers. After-Call-Work (ACW) is the average amount of time an agent takes to wrap-up a call. Average Handle Time is the combination of both ATT and ACW.
Calculation: Average-Talk-Time + After-Call-Work
Contact centres tend to place tremendous effort in driving down AHT because it is directly related to increased cost. In very large centres, an additional second of AHT can add as much as one million dollars in additional annualized costs. It is important however not just to monitor AHT but also the effectiveness of a call. The key is to find the right balance for your business.
Example:
Agent 'A' spends 190 seconds in ATT and completes all the necessary administrative duties (such as completing an order or opening a service ticket) while the customer is on the phone and spends only 20 seconds in ACW.

Agent 'B' spends 140 seconds speaking to the customer, hangs up and spends 70 seconds in ACW completing necessary administrative duties.

Both agents have an equal AHT of 210 seconds, from a cost perspective they have performed equally. The question now is which customer ended the call feeling most satisfied.

### Service Level & Average Speed of Answer (ASA)
Service level is usually defined as the percentage of calls answered within a predetermined number of seconds. If your service level target is 80/20, then you are striving to answer 80% of all calls within 20 seconds (or about 5-6 "rings"). The faster an agent answers a customer call, the higher service levels tend to be. Once your "seconds" target is established, you can report your service level as a percentage. There is no defined standard but many contact centres strive for 80/20 or 80/30 as their standard for tracking this metric. Answering calls too quickly can result in significant additional costs to the business (because of the significant labour costs required to meet very high service levels). Some centres are able to operate at 70/40 based on business needs and client expectations. It is important that this metric remain fluid, based on customer type, call type, campaign, and competitive landscape.
Average Speed of Answer (commonly referred to as ASA) is the average number of seconds it takes for a call to be answered. If calls are, on average, answered in 15 seconds, the ASA is 15 seconds for the contact centre.
Generally, the shorter the ASA, the better your Service Level and vice versa.
### First Call Resolution
First Call Resolution (FCR) is a relatively new metric to the contact centre industry. FCR measures the percentage of customer issued calls that are resolved the first time.
Calculation: Number of FCR calls / Total Number of Calls
The difficulty is defining what a successful FCR is when millions of calls are arriving in your contact centre. It is a challenge to effectively track whether a customer has received a resolution they are satisfied with.
Contact centres will use a time frame within the first customer call to determine if that call was successfully resolved in the first call. For example some centres will use a 72 hour time frame to help determine a negative or positive FCR. If a customer calls back within a 72-hour period then this is considered an unsuccessful FCR. There are inherent flaws in this calculation – for instance when a customer calls back after 72 hours with the same issue or within 72 hours with a new or different issue. Some contact centres use customer satisfaction surveys to gain further insight into FCR.
Although there are flaws with this type of tracking, monitoring the average of FCR will give you some sense as to how well your agents are able to resolve a customer’s issue on the first call.
In a time when the customer experience is paramount, FCR is a key metric to measure and understand. Some contact centres will say this is the most important metric to analyze, especially when you consider that a Customer may wait a while to reach an agent and receive extreme satisfaction with the call. This metric is also important to consider when it is compared to a Customer who reaches an agent immediately but the agent is unable to help the Customer.
A very high FCR can save your organization considerable money (the result is less repeat calls) and greatly improves customer satisfaction.
Example:
If 8 of 10 calls are resolved on the first call then that centre has reached an FCR of 80%.
7. Occupancy
Occupancy defines how well staff is scheduled for the call volumes coming into your contact centre. An inblockquoteidual’s occupancy rate can also be determined by using the same formula.
Calculation: Total Call Time / (Total Call Time + Available Time)
There is no call centre standard for occupancy but if you have low occupancy (i.e. <70%) you risk boredom by your agents. If you have consistently high occupancy (i.e. >85%) you risk over-extending your agents which can lead to higher and faster turnover. Many contact centres try to keep their occupancy metric between 70% and 80% to maintain a healthy balance.
Example:
If your contact centre has 100 hours of time your agents can take calls, and they are speaking to Customers for 80 hours of that 100 hours, your occupancy is 80/100 or 80%.
8. Cost per Call (Contact)
Cost per call is calculated as the cost of a call arriving and being answered at your contact centre. There are many ways to calculate this, and here we will focus on the basic cost per call and the fully loaded cost per call.
Basic Cost per Call:
Calculation: Calls per Hour / Agent’s Hourly Wage
Basic cost per call is useful when determining the extra cost associated with handling a nominal number of additional calls. Once you reach a certain level of increased calls you may need to add additional agents, supervisors and stations.
Example:
If your agent takes 15 calls per hour and is paid an hourly wage of $15 then the cost per call is $1. In order to handle an additional 100 calls within your contact centre you will have to pay $100 more.
Fully Loaded Cost per Call:
Calculation: (Annual Operating Cost + Labour) / Total Number of Calls per Year
Annual operating costs take into account facilities, supervisors, managers, benefits, incentives, maintenance, desktops, etc. Fully loaded cost per call is useful if you want to compare your efficiencies with that of a third party contact centre operation. Most contact centres’ budgeted cost per call metrics are based on a fully loaded cost.
Example:
If your annual operating costs are $100,000 per year and your labours costs are $1,000,000 and you receive 4,000,000 calls per year the fully loaded cost per call is $.28
9. Turnover (Attrition)
Turnover (also known as attrition) is a commonly tracked metric in the contact centre industry.
There are two types turnover:
1. Voluntary - Staff choosing to leave
2. Involuntary - Staff being asked to leave
Turnover measures the number of people who leave your contact centre (as a percentage).
Calculation: Number of People to Leave / Number of Positions
Turnover can be a positive or a negative. Some turnover due to performance management (terminating a poor performer) is a positive for the contact centre. If good performers leave, you have turnover that would be considered negative in your contact centre. You should also consider whether people are leaving your contact centre for other positions within the company which would be considered good turnover from an organization’s perspective.
Turnover varies depending on what your contact centre is accomplishing. High turnover in 3rd party outbound contact centres is common, since staffing is constantly adjusted for programs that are starting and finishing. In an in-house contact centre where call volumes are predictable and steady, your turnover should be lower.
High turnover can be expensive for two reasons.
Staff will be relatively untrained which can lead to problems in dealing with issues consistently. It costs to train agents and during training they are not contributing to productivity.
Increased recruitment and training costs, poor Customer experience, potential lost sales opportunities, low morale if you are understaffed for your call volume and agents may see the call centre as a “revolving door”.
Low turnover can be both a benefit and a liability. The benefit is that you have seasoned agents who understand the business and the customers. The liability can be in the cost of staffing as vacation time grows for more tenured agents. A 20 year veteran team of agents may receive a relatively high hourly pay rate and 6 or more weeks of vacation time. This can pose a challenge to costs and staffing.
Some call centres are analyzing voluntary turnover to understand why key staff are leaving the business or if the departing staff are employees you wanted to leave. If you are losing your key people this could be a sign of a bigger problem within the organization, such as poor morale.
High involuntary turnover may be a sign of poor hiring practices. In this case new hires may not be appropriately matched for the roles and the culture of the organization.
Example:
If there 100 full-time staff positions and 12 people leave in a year, your average turnover is 12% for that year (1 person leaves, on average, each month).
10. Call Quality
Call Quality is a standard scoring/rating system that contact centres use to determine how well an agent deals with the customers. There are no industry standards for monitoring quality, but there will usually be a list of criteria that an agent must cover during a call. This includes, but is not limited to, how the agent answers the call, how they navigate the caller to a resolution, and how they end the call.
Calculation: Number of Criteria Met / Number of Total Criteria
Management can have ten or more criteria for what they consider to be a good call. Some of the criteria are weighted based on importance (eg, effectively probing to understand client’s needs). Usually the criteria are added up to an ultimate quality score that is displayed as a percentage. The benefit to breaking down the call into inblockquoteidual criteria is to have specific areas of development to coach agents and improve call quality.
Many contact centres are adding FCR to their scoring criteria to focus on the customer experience. Some contact centers look at both call quality scores as well as scores achieved from the customer satisfaction survey. An agent may score 70% in call quality and a 90% on the customer satisfaction survey. In this regard, contact centres must consistently balance quality scores and customer experience.
Quality assessments are not limited to calls. Contact centres should score e-mail and chat functions for quality as well as any other customer interactions.
Example:
If a contact centre has 20 criteria for a perfect call and the agent executes against 18 of the criteria well then the agent’s score is 90%. The overall quality score for the contact centre is the weighted average of all agents’ quality scores. The number of calls monitored to determine a quality score varies from centre to centre, although between 5 to 15 calls monitored and scored per agent per month is common.